Explaining controllers as viable alternative

ThomasLandauer opened the pull request:
See https://github.com/symfony/symfony-docs/issues/13406#issuecomment-614697619

Please wrap it in a "caution" box, or whatever you think is appropriate.

I'm unsure about how much context of the controller should be shown in the code, since the entire controller is only shown later. If you think it's too confusing now, I'd suggest to show the controller as "skeleton" above this note.

ThomasLandauer committed:
Explaining controllers as viable alternative

See https://github.com/symfony/symfony-docs/issues/13406#issuecomment-614697619

ThomasLandauer committed:
Update voters.rst

Fixing https://github.com/symfony/symfony-docs/pull/13554/checks?check_run_id=592603240

wouterj committed:
[#13554] Slightly reworded the tip

wouterj reviewed (approved):
Hi @ThomasLandauer! I've added a commit to this PR to transform it into a tip & slightly reworded the message. Please let me know if you disagree with any changes

ThomasLandauer committed:
Update voters.rst

ThomasLandauer commented:
> *slightly* reworded

lol

I re-added the first sentence (plus an additional "instead"). We need to emphasize that this is not just some "introduction" to voters, but rather a completely different approach. And the first sentence somehow tries to explain that the following example isn't a good real-world example for the usage of voters. I thought about adding something like "for simplicity's sake", but couldn't find a good solution...

noniagriconomie commented on security/voters.rst:
```diff
@@ -31,6 +31,22 @@ in the application, which can be: affirmative, consensus or unanimous.
+
+    The example used throughout this page features just two routes (``post_show`` and ``post_edit``).
+    However, the advantage of voters is that you can reuse them in *many* places and centralize
+    all permission logic. If you don't reuse permissions and the rules are simple, you instead
```
```suggestion
    all permission logic. If you don't reuse permissions or the rules are basic, you instead
```

wouterj commented:
> And the first sentence somehow tries to explain that the following example isn't a good real-world example for the usage of voters.

I think it is. I wouldn't call the permission logic in this example "simple"/"basic" anymore (75+ lines of code). I don't think the first sentence adds much to the tip (it's really a separate tip) and I wanted to keep it as small as possible (long side blocks are often distracting), that's why I removed it.

ThomasLandauer committed:
Update security/voters.rst

Co-Authored-By: Antoine Makdessi <antoine.makdessi@agriconomie.com>

ThomasLandauer commented:
What I'm trying to say: The following example is "basic" in the sense that it fulfills exactly our "new" criteria for doing it differently:

> If you don't reuse permissions or the rules are basic, you instead might want to

There's no reuse, and the rules are simple. So my full message would read:

> If your app is exactly like ours here, don't use voters. We chose this unrealistically simplified example in order to be able to show you how voters work in just 75+ lines, not 200+.

Don't know how to get this across in just one short sentence ;-)

weaverryan commented on security/voters.rst:
```diff
@@ -31,6 +31,22 @@ in the application, which can be: affirmative, consensus or unanimous.
+
+        // ...
+        if ($post->getOwner() !== $this->getUser()) {
+            throw new AccessDeniedException();
```
Let's use: `throw $this-> createAccessDeniedException()` - we always assume users are using the normal shortcuts - like `AbstractController`.

ThomasLandauer replied:
Done :-)

weaverryan commented on security/voters.rst:
```diff
@@ -31,6 +31,22 @@ in the application, which can be: affirmative, consensus or unanimous.
+    However, the advantage of voters is that you can reuse them in *many* places and centralize
+    all permission logic. If you don't reuse permissions or the rules are basic, you instead
+    might want to do the check in the controller directly and throw an ``AccessDeniedException``
+    to create the correct response::
```
I'd say we move this to the VERY top of this chapter - like the very first words. Then we can say something simpler:

```rst
If you need to deny access in a controller based on custom rules - perhaps only the "owner" of an object can access a page - you can always put that logic directly in your controller::

    ... the code

But if you need to repeat this logic in multiple places, a better option is to centralize that logic with a custom voter.
```

And then it would go into the current beginning "In order to use voters".

I think you're right to talk about this - so why not discuss the easy option (and when it might not work) immediately? :)

ThomasLandauer replied:
I tried my best :-)

ThomasLandauer committed:
Update voters.rst

See https://github.com/symfony/symfony-docs/pull/13554#discussion_r410150210

ThomasLandauer committed:
Update voters.rst

Trying to merge the suggestions of 4 people ;-)

What's really odd is that on this dedicated voters page, the reader is referred to someplace else for more info on voters ;-)

> Take a look at the authorization article for an even deeper understanding on voters.

So in the long run, these two should be merged - or at least the voters part integrated into this page.

xabbuh commented on security/voters.rst:
```diff
@@ -6,19 +6,30 @@
+        throw $this->createAccessDeniedException();
+    }
+
+In that sense, the following example used throughout this page is more like a
```
I am fine with adding the explanation above, but I do not agree with this sentence. The example used below IS a real-world use case. I suggest to remove this sentence.

ThomasLandauer replied:
I removed the second part - please take a look again :-)

ThomasLandauer committed:
Update voters.rst

See https://github.com/symfony/symfony-docs/pull/13554#discussion_r411850997

xabbuh commented:
Thank you @ThomasLandauer.

ThomasLandauer commented:
@xabbuh It looks like this wasn't up-migrated to 4.4, 5.1 etc:
* It's there at https://symfony.com/doc/3.4/security/voters.html
* It's not there at https://symfony.com/doc/4.4/security/voters.html

Should I commit it to 4.4 too?

xabbuh commented:
There were some conflicts when merging branches and I didn't have the time to look at them. Will probably manage to do so tomorrow.
